### PR TITLE
Polish pull request details view

### PR DESCRIPTION
**Pull request by donokuda**
Fixes https://github.com/github/VisualStudio/issues/709 https://github.com/github/VisualStudio/issues/713 https://github.com/github/VisualStudio/issues/625 https://github.com/github/VisualStudio/issues/302

This pull request reworks the details view. I reworked around the information so that it's quicker to parse than before.

**Before**

screen shot 2016-12-12 at 3 48 31 pm

**After**

screen shot 2016-12-12 at 2 44 50 pm

**Push / Pull branch actions**

screen shot 2016-12-12 at 3 53 59 pm

**Shorter title and checkout branch action**

screen shot 2016-12-12 at 3 54 19 pm

**Themes!**

screen shot 2016-12-12 at 2 44 35 pm
screen shot 2016-12-12 at 2 44 14 pm

### TODO

- [x] Dark theme
- [x] Light theme
- [x] Don't hardcode font sizes
- [x] Double check design for error messages
- [ ] Handle long branch names (wrap or truncate perhaps?)

### CODE REVIEW

**Comment by saraford** [comment]
The 2nd image implies that the name of the local repo will always match the remote repo `commits-behind` (since I would click on `checkout commits-behind` in the previous pane). This got me thinking... is there any way for users to get to this pane where the local branch has a different name than the remote? E.g. is it possible to rename this local branch in TE? If so, what would happen if the user tried to pull or push?

**Comment by saraford** [comment]
another thought... at the top we show the user the PR number and Repo. Would showing the current local branch also be helpful? Or does that become too wordy?

**Comment by grokys** [comment]
> The 2nd image implies that the name of the local repo will always match the remote repo commits-behind (since I would click on checkout commits-behind in the previous pane).

`commits-behind` here is the name of the branch, not the repository.

> is there any way for users to get to this pane where the local branch has a different name than the remote? E.g. is it possible to rename this local branch in TE? If so, what would happen if the user tried to pull or push?

Yes, it is possible to rename a branch in TE. In the case of a PR from the same repository, this is no different from checking out the PR branch normally (i.e. not using the PR pane) and renaming it however. All we're doing in that case is checking out a branch normally. In the case of a branch checked out from a fork, then the push and pull should still push and pull to the correct branch on the fork,

> Would showing the current local branch also be helpful? Or does that become too wordy?

The current branch is always shown in the status bar at the bottom right, so I'm not sure that adding it here would add any extra info.

File: src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml [hunk 1]
```diff
@@ -113,108 +120,118 @@
...
+
+                    <!-- Pull/push buttons -->
+                    <StackPanel Orientation="Horizontal" VerticalAlignment="Center"
+                                Visibility="{Binding UpdateState.UpToDate, FallbackValue=Collapsed, Converter={ui:BooleanToInverseVisibilityConverter}}">
+                        <ui:OcticonImage Icon="arrow_down"/>
+                        <TextBlock Text="{Binding UpdateState.CommitsBehind}" VerticalAlignment="Center"/>
+                        <ui:GitHubActionLink Content="Pull"
+                                             Command="{Binding Pull}"
+                                             Margin="4,0"
+                                             ToolTip="{Binding UpdateState.PullDisabledMessage}"
```
Comment (donokuda) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: Would it make sense to have a tooltip that says something like "Pull changes from the remote branch foo-branch to your local branch?" If so what's the best way to do this so that it also works with this disabled message?

File: src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml [hunk 2]
```diff
@@ -113,108 +120,118 @@
...
+                    </TextBlock>
+
+                    <!-- Git operation error message -->
+                    <TextBox Foreground="Red" 
+                             Margin="0 4"
+                             Text="{Binding OperationError}"
+                             VerticalAlignment="Center"
+                             TextWrapping="Wrap"
+                             Style="{StaticResource FlatReadOnlyTextBox}"
+                             Visibility="{Binding OperationError, Converter={ui:NullToVisibilityConverter}}"/>
```
Comment (donokuda) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: What's the best way to test this? Should I set the `Visibility` to something like `true`?

Comment (grokys) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: Yeah you can do that, or you can set the `OperationError` property on the sample data.

File: src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml (hunk 1 above)
Comment (grokys) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: Hmm, that will require a few changes. Probably best to merge this without them an add them after?

Comment (donokuda) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: Yup, that's a good call 👍

File: src/GitHub.VisualStudio.UI/Styles/ThemeDesignTime.xaml [hunk 3]
```diff
@@ -4,7 +4,7 @@
   xmlns:PresentationOptions="http://schemas.microsoft.com/winfx/2006/xaml/presentation/options">
 
   <ResourceDictionary.MergedDictionaries>
-    <ResourceDictionary Source="pack://application:,,,/GitHub.VisualStudio.UI;component/Styles/ThemeBlue.xaml" />
+    <ResourceDictionary Source="pack://application:,,,/GitHub.VisualStudio.UI;component/Styles/ThemeLight.xaml" />
```
Comment (grokys) [inline src/GitHub.VisualStudio.UI/Styles/ThemeDesignTime.xaml]: Did you mean to commit this change or was this part of you checking other themes?

**Comment by grokys** [comment]
LGTM @donokuda! I commited a small change to remove a comment that didn't seem necessary any more and asked one question. Once you've answered the question, looks good to merge!

File: src/GitHub.VisualStudio.UI/Styles/ThemeDesignTime.xaml (hunk 3 above)
Comment (donokuda) [inline src/GitHub.VisualStudio.UI/Styles/ThemeDesignTime.xaml]: This was me checking out other themes. I'll change this back to make the diff look better by one line 😉

File: src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml (hunk 2 above)
Comment (donokuda) [inline src/GitHub.VisualStudio/UI/Views/PullRequestDetailView.xaml]: Awesome. I'll check this out and then merge if it looks good.

**Comment by donokuda** [comment]
@grokys I reworked the `<StackPanel>` to be a `<Grid>` instead so that error messages don't get cut off. I wasn't able to test it in the wild (not exactly sure how to, but I'll take a pass if someone points me in the right direction.)

Let me know how this looks and if it's good to merge still. 🙌

**Comment by grokys** [comment]
Are we ready for a merge? There's still "Double check design for error messages" left unchecked.

**Comment by shana** [comment]
I'm really late to this party but just wanted to note that I think this design is really nice and readable, looks great! :heart:

**Comment by donokuda** [comment]
@grokys I just doubled checked and looks like I need to tidy up this example of modern art 😄 

image

Throwing back on the WIP label but will 🏓 when it's ready.

**Comment by grokys** [comment]
It's a masterpiece! ;)

I also noticed that the "`foo` into `bar`" message gets cut off if the branch names are long. Don't know if you wanted that to wrap?

image

**Comment by donokuda** [comment]
hah! more like "`foo` into the margin of the tool window!" I'll try a version of it wrapping. If it feel awkward, then I'll try a pass with truncating.

Good looking out 👀

**Comment by donokuda** [comment]
Error messages now look like this:

screen shot 2016-12-16 at 10 33 40 am

I'm also trying to figure out the best way to handle long branch names. I moved everything to a `<Grid>` so that the width of the branch name adjusts based on the available name. This works fine when branch names (specifically the source branch) is long:

screen shot 2016-12-16 at 4 28 38 pm

But it falls apart when the branch name is short 😢:

screen shot 2016-12-16 at 2 28 14 pm

Ideally, what I would like it to do is have the `MaxWidth` of the column with the source branch name to be the equivalent of `Width="*"` and the `Width` to be `Auto` (if that makes any sense.) I'm stumped on how to do this... 😕 

I can haz halp? 🙏

**Comment by grokys** [comment]
Just spoke with @shana and we'd like to get a beta prepared so going to merge this as-is and fix the spacing after.

**Comment by donokuda** [comment]
@grokys Sounds good. Thanks for pushing this one over the finish line! 🙇
